### PR TITLE
ECal Portion of DR Geometry Update

### LINKED ISSUE #890: More Realistic ECal Simulation
After attending the recent ECAL meeting, I figured there are a few changes we could make to the `ecal.gdml` that would improve the simulation's realism ~~without having to change any other objects in our software~~ (see comment below).

These changes are summaries from [a presentation](https://github.com/LDMX-Software/ldmx-sw/files/5110834/LDMX_layout.pdf) given by Susanne Kyre (UCSB) at an [ECal Meeting](https://indico.fnal.gov/event/45052/). 

## Rigid Changes that Definitely should be put in
- [x] Tungsten layers aren't seven individual hexagons but one large octagon
- [x] Carbon cooling planes stretch to the edge of the ECal box and are not just hexagons

## Changes that Haven't been decided upon, so maybe we shouldn't implement them
- [x] Motherboards are not big rectangles but are much more complicated objects, not sure on their shape
- [x] 5 ROCs per sensitive layer each ~ 8cm x 8cm with a ~3cm overlap with Motherboards for mounting, not sure on their placement
- [x] Getting Tungsten thickness to be less than 1mm is technologically challenging and expensive, might want to change the absorber thickness in group A to 1mm rather than 0.75mm.

One of the images from Susanne Kyre's presentation is helpful to see what my descriptions mean. The red hexagons are the sensitive modules, the tan octogon is the tungsten layer, the grey rectangle is the carbon cooling plane, the bright green are the motherboards, and the dark green are the ROCs.
ecal_layout

### LINKED ISSUE #904: ECal Layer Shifting and incorporate layerID in ECalHexReadout
**Is your feature request related to a problem? Please describe.**
Layer-shift the second layer within each bi-layer of Ecal layers--- to cover cracks.
ECalHexReadout is a x-y map of cells, but does not describe how x and y changes with z.
ECalID.h has layers defined in terms of double layers (i.e. 17 layers with 12 ring modules in a layer)--- not convenient for layer shifting (need to separately define 34 layers)

**Describe the solution you'd like**
Need to add layerID and incorporate z as a function of x-y in ECalHexReadout.h and .cxx.
Add Layer shifting in ECalHexReadout.
Change ECalID so that there are 34 layers and 6 ring modules in each.

**Describe alternatives you've considered**
Link the ecal.gdml file to ECalHexReadout so that it's not necessary to change ECalHexReadout every time a new layer-configuration is tried out?

**Additional context**
layer-shift
One module layer-shift

### LINKED ISSUE #923: Rename and Clean-Up EcalHexReadout
EcalHexReadout is a class that we are now using for almost all translation between EcalID and real-space position, not just cutting up the module hexagons into cellular hexagons. With this in mind, it makes a lot more sense to call it something more general than `EcalHexReadout`, probably `EcalGeometry` would work best, and then it would line up with the provider that provides it `EcalGeometryProvider`.

I'm also going to take this opportunity to clean-up the Ecal precision digis and the HgcrocEmulator. I need to make the collection name and pass name for the inputs configurable.

@omar-moreno I would love to put all of this Ecal-related stuff in the Ecal module. After the cmake changes you are working on, how difficult would it be to make a "Geometry" sub-module of Ecal (similar to how we are making "Event" submodules)? This way, `SimCore` and `Ecal::Ecal` could link to `Ecal::Geometry` and the source for `Ecal::Geometry` would be in the `Ecal` module.

### PR DESCRIPTION
**Pull request by tomeichlersmith**
I am updating _ldmx-sw_, here are the details.

**This is a draft PR because I have not tested the GDML updates in simulation; however, I think I've implemented the complicated C++ side of things inside of `EcalGeometry` (formerly `EcalHexReadout`) and the visualization of the GDML detector looks correct.**

### What are the issues that this addresses?
- resolves #904
- resolves #923 
- resolves #890 

## Check List
- [x] I successfully compiled _ldmx-sw_ with my developments
- [x] I ran my developments and the following shows that they are successful.
- [x] I attached any sub-module related changes to this PR.

### Related Sub-Module PRs
- https://github.com/LDMX-Software/Ecal/pull/41
- https://github.com/LDMX-Software/SimCore/pull/71

### CODE REVIEW

**Comment by tomeichlersmith** [comment]
I'm using the python script below to test that the geometry can catch all hits throughout the entire ECal flower.

```python
from LDMX.Framework import ldmxcfg
p = ldmxcfg.Process( 'test_geometry' )
p.maxEvents = 1000
p.run = 0
p.outputFiles = ['test_geometry.root']
import LDMX.Hcal.HcalGeometry
import LDMX.Ecal.EcalGeometry
from LDMX.SimCore import simulator
from LDMX.SimCore import generators
sim = simulator.simulator("single_muons")
import sys
sim.setDetector(sys.argv[1] , True)
sim.description = "ECal muons"
gpsCmds=[ "/gps/particle mu-",
          "/gps/pos/type Plane",
          "/gps/direction 0 0 1",
          "/gps/ene/mono 4 GeV", # fixed energy
          "/gps/pos/shape Square",
          "/gps/pos/centre 0 0 220. mm", # start right in front of ECal
          "/gps/pos/halfx 400 mm",
          "/gps/pos/halfy 300 mm",
          "/gps/number 1",
          ]
myGPS = generators.gps( 'myGPS' , gpsCmds )
sim.generators.append(myGPS)
p.sequence = [ sim ]
```
This is derived from Hcal/test/hcal_test_geometry_config.py

File: DetDescr/include/DetDescr/EcalGeometry.h [hunk 1]
```diff
@@ -296,7 +194,7 @@ class EcalHexReadout : public framework::ConditionsObject {
    */
   std::vector<EcalID> getNN(EcalID id) const {
     auto list = NNMap_.at(EcalID(0, id.module(), id.cell()));
-    for (auto& flat : list) id = EcalID(id.layer(), flat.module(), flat.cell());
+    for (auto& flat : list) flat = EcalID(id.layer(), flat.module(), flat.cell());
     return list;
   }
```
Comment (tomeichlersmith) [inline DetDescr/include/DetDescr/EcalGeometry.h:215]: This patch is what is causing the deviation of Summed Iso between this branch and trunk.

File: DetDescr/include/DetDescr/EcalGeometry.h [hunk 2]
```diff
@@ -322,7 +220,7 @@ class EcalHexReadout : public framework::ConditionsObject {
    */
   std::vector<EcalID> getNNN(EcalID id) const {
     auto list = NNNMap_.at(EcalID(0, id.module(), id.cell()));
-    for (auto& flat : list) id = EcalID(id.layer(), flat.module(), flat.cell());
+    for (auto& flat : list) flat = EcalID(id.layer(), flat.module(), flat.cell());
     return list;
   }
```
Comment (tomeichlersmith) [inline DetDescr/include/DetDescr/EcalGeometry.h:241]: ... and this patch

**Comment by tomeichlersmith** [comment]
The PR Validation plots which use the v12 geometry _failed_ due to a single plot failing its KS-test in the it_pileup, ecal_pn, and inclusive validation samples. This plot was the "ECal Summed Iso" plot which plots the total energy in the ECal that is from hits that do not have hits in nearest neighbor or next-nearest neighbor cells. (Plots extracted and attached).
[it_pileup_summed_iso.pdf](https://github.com/LDMX-Software/ldmx-sw/files/9382533/it_pileup_summed_iso.pdf)
[ecal_pn_summed_iso.pdf](https://github.com/LDMX-Software/ldmx-sw/files/9382534/ecal_pn_summed_iso.pdf)
[inclusive_summed_iso.pdf](https://github.com/LDMX-Software/ldmx-sw/files/9382535/inclusive_summed_iso.pdf)

### Why
This shift up in slightly more isolated energy across all of these validation samples is due to a patched-bug that I discovered while reworking EcalGeometry (formerly EcalHexReadout). I commented on this bug above.

Basically, EcalHexReadout on trunk _does not_ insert the layer number into the EcalIDs that are labeled as NN or NNN, so the summed iso is only catching energy that is isolated in the zero'th layer.

**Comment by tomeichlersmith** [comment]
Running the above geo-test config over v12 and v14 detectors now is functional :+1: 

## Still To Do
- [x] First version of layer weights into ECal
- [x] Check if updated scoring planes overlap with ECal anywhere
- [ ] Add some "fins" to ECal parent volume to encompass "fins" of support box

**Comment by tomeichlersmith** [comment]
## Final Testing
I generated 1M geantino events to really blanket the ECal to try to make sure all nooks-and-cranies were checked. The sim successfully ran and I double-checked that the output file did indeed contain ecal sim hits throughout the detector.

Upon further reflection, I am choosing to **not** implement the encompassing fins that are necessary to completely enclose the suport box in the ecal parent volume. This means that overlap checking within ecal.gdml should return one (and only one) failure pointing out that the ecal support box and the parent volume `ECal_box` overlap. Then we will need to check that the fins do not overlap with anything else manually. This is easily done in the visualization.

The small benefits from enclosing the fins is outweighed by the hassle of trying to figure out how to enclose them (for now). Perhaps if further annoyances arise, I'll look into trapezoids more.

## geantino sim hits
ecal_sim_hits_x
ecal_sim_hits_y
ecal_sim_hits_z

## geantino config
```python
from LDMX.Framework import ldmxcfg
p = ldmxcfg.Process( 'test_geometry' )
p.maxEvents = 1000000
p.run = 0
p.termLogLevel = 0
p.outputFiles = ['test_geometry.root']
import LDMX.Hcal.HcalGeometry
import LDMX.Ecal.EcalGeometry
from LDMX.SimCore import simulator
from LDMX.SimCore import generators
sim = simulator.simulator("single_muons")
import sys
detname = 'ldmx-det-v14'
if len(sys.argv) > 1 :
    detname = sys.argv[1]
sim.setDetector(detname , True)
sim.description = "ECal geantino geometry test"
sim.verbosity = 5
gpsCmds=[ "/gps/particle geantino",
          "/gps/pos/type Plane",
          "/gps/direction 0 0 1",
          "/gps/ene/mono 4 GeV", # fixed energy
          "/gps/pos/shape Square",
          "/gps/pos/centre 0 0 240. mm", # start right in front of ECal
          "/gps/pos/halfx 420 mm",
          "/gps/pos/halfy 300 mm",
          "/gps/number 1",
          ]
myGPS = generators.gps( 'myGPS' , gpsCmds )
sim.generators.append(myGPS)
p.sequence = [ sim ]
```

**Comment by tomeichlersmith** [comment]
The PR Validation plots failed in the same locations as [the first round of checks](https://github.com/LDMX-Software/ldmx-sw/pull/1079#issuecomment-1220691835) so we are ready to merge.
